[AArch64][ELF] Section alignment of 4 for AArch64 instruction

popaflorin opened the pull request:
The integrated assembler sets a minimum alignment for the .text section of 4. However user defined sections get an alignment of 1. Unlike the GNU assembler which raises the section alignment to 4 if an AArch64 instruction is used, the integrated assembler leaves the alignment at 1

flopop01 committed:
[AArch64][ELF] Section alignment of 4 for AArch64 instruction

The integrated assembler sets a minimum alignment for the .text section of 4. However user defined sections get an alignment of 1. Unlike the GNU assembler which raises the section alignment to 4 if an AArch64 instruction is used, the integrated assembler leaves the alignment at 1

flopop01 committed:
[AArch64][ELF] Section alignment of 4 for AArch64 instruction

The integrated assembler sets a minimum alignment for the .text section of 4. However user defined sections get an alignment of 1. Unlike the GNU assembler which raises the section alignment to 4 if an AArch64 instruction is used, the integrated assembler leaves the alignment at 1

flopop01 committed:
[AArch64][ELF] Section alignment of 4 for AArch64 instruction

The integrated assembler sets a minimum alignment for the .text section of 4. However user defined sections get an alignment of 1. Unlike the GNU assembler which raises the section alignment to 4 if an AArch64 instruction is used, the integrated assembler leaves the alignment at 1

flopop01 committed:
[AArch64][ELF] Section alignment of 4 for AArch64 instruction

The integrated assembler sets a minimum alignment for the .text section of 4. However user defined sections get an alignment of 1. Unlike the GNU assembler which raises the section alignment to 4 if an AArch64 instruction is used, the integrated assembler leaves the alignment at 1

flopop01 committed:
[AArch64][ELF] Section alignment of 4 for AArch64 instruction

The integrated assembler sets a minimum alignment for the .text section of 4. However user defined sections get an alignment of 1. Unlike the GNU assembler which raises the section alignment to 4 if an AArch64 instruction is used, the integrated assembler leaves the alignment at 1

lenary commented on llvm/lib/MC/MCParser/ELFAsmParser.cpp:
```diff
@@ -697,6 +697,15 @@ bool ELFAsmParser::ParseSectionArguments(bool IsPush, SMLoc loc) {
+
+  // Section alignment of 4 if an AArch64 instruction is used when $x mapping
+  // symbol is added Match GNU Assembler
+  const Triple &TT = getContext().getTargetTriple();
+  if ((Section->getFlags() & ELF::SHF_EXECINSTR) && (TT.isAArch64())) {
+    if (Section->getAlign() < 4)
+      getStreamer().emitValueToAlignment(Align(4));
+  }
```
You don't need to do this in target-independent code - `MCStreamer::switchSection` will call `MCStreamer::changeSection` (the latter is `virtual`, and `AArch64ELFStreamer` has an override of it already, where you can add logic like this (Note this code is also handling the mapping symbols):

 https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AArch64/MCTargetDesc/AArch64ELFStreamer.cpp#L191-L205

popaflorin replied:
I have moved the logic into `MCStreamer::changeSection` of the AArch64ELFStreamer as suggested

Stylie777 commented on llvm/test/MC/AArch64/directive-arch-section-alignment.s:
```diff
@@ -0,0 +1,21 @@
+// CHECK-NEXT: 4: d503201f      nop
+// CHECK-NEXT: 8: d503201f      nop
+// CHECK: 0000000000000000 <sec01>:
+// CHECK-NEXT: 0: d503201f      nop
```
How are we checking the alignment here to ensure it has been set to 4, especially in the `sec00` section?

popaflorin replied:
I am not sure how to check that. I thought if the address is 4 bytes increment, that will imply an alignment of 4? But I don't know yet how to do it better because since I've changed the code, it doesn't emit `p2align` directive anymore

Stylie777 replied:
If you add the `-h` option to the `llvm-objdump` that should give you the section header information, which should include the alignment.

popaflorin replied:
I have redone the test and now I am checking for the section alignment with both `readobj` and 'readelf'

Stylie777 commented on llvm/lib/Target/AArch64/MCTargetDesc/AArch64ELFStreamer.cpp:
```diff
@@ -202,6 +202,12 @@ class AArch64ELFStreamer : public MCELFStreamer {
     MCELFStreamer::changeSection(Section, Subsection);
+
+    // Section alignment of 4 to match GNU Assembler
+    if (Section->getAlign() < 4) {
```
Yes - this should be updated to be like before @popaflorin where you were checking for if AArch64 instructions were used.

popaflorin replied:
Thank you both for your feedback! I haven't figured out how to do that yet in this AArch64ELFStreamer. I thought this will suffice: `Section->hasInstructions()` for a check but it always comes back as 0. I'm looking into it...

lenary replied:
it might be better/easier to check if the section is executable? that's probably the only info you have when the section is brand new (I think `hasInstructions` only becomes true when the first instruction is emitted into the section)

popaflorin replied:
I am checking now `Section->isText()` in the latest patchset which seems to get the correct executable section. In the test I have also included a `aw` section to prove that the alignment will remain unchanged at 1

popaflorin commented on llvm/test/MC/AArch64/directive-arch-section-alignment.s:
```diff
@@ -0,0 +1,21 @@
+// RUN: llvm-objdump -d -r %t.obj | FileCheck %s
+
+.section sec00, "ax"
+nop
```
I have redone the test where I include various sections examples and check for the alignment with `readobj` and `readelf`

lenary commented on llvm/lib/Target/AArch64/MCTargetDesc/AArch64ELFStreamer.cpp:
```diff
@@ -202,6 +202,12 @@ class AArch64ELFStreamer : public MCELFStreamer {
+
+    // Section alignment of 4 to match GNU Assembler
+    if ((Section->getAlign() < 4) && Section->isText()) {
+      Section->setAlignment(Align(4));
+      emitValueToAlignment(Align(4), 0, 1, 0);
```
Can you clarify why you need both the `setAlignment` on the section, and the `emitValueToAlignment` (which will insert bytes until the offset in the section is so aligned)

Am I missing when the latter is needed?

popaflorin replied:
This was a leftover from initial testing and I forgot to remove it.  I have now corrected it in the latest patch

lenary commented on llvm/lib/Target/AArch64/MCTargetDesc/AArch64ELFStreamer.cpp:
```diff
@@ -202,6 +202,11 @@ class AArch64ELFStreamer : public MCELFStreamer {
     MCELFStreamer::changeSection(Section, Subsection);
+
+    // Section alignment of 4 to match GNU Assembler
+    if ((Section->getAlign() < 4) && Section->isText()) {
+      Section->setAlignment(Align(4));
+    }
```
```suggestion
    if (Section->isText())
      Section->ensureMinAlignment(Align(4));
```

lenary reviewed (commented):
Since the most recent update, `llvm/test/MC/AArch64/directive-arch-section-alignment.s` seems to be failing on Linux at least - can you check it again please?

I also have a suggestion, both for clarity (using ensureMinAlignment) and code style (no braces on a single-statement if).

lenary reviewed (approved):
LGTM! Thanks!

lenary commented:
Rebasing the patch, there seems to be a problem with `lld/test/ELF/aarch64-relocs.s` (this would have also happened after a merge). Can you check that testcase? I think the problem is likely the testcase itself - but I'm not entirely sure. You'll need to build LLD, which shouldn't add much overhead to your existing LLVM build.

Stylie777 reviewed (approved):
Changes LGTM! Thanks @popaflorin

flopop01 committed:
[AArch64][ELF] Updated the synthax and fixed the lld test

As a result of review feedback, updated the synthax for the change and
first attempt to fix lld test that was failing due to missalignment

flopop01 committed:
[AArch64][ELF] Updated the synthax and fixed the lld test

As a result of review feedback, updated the synthax for the change and
first attempt to fix lld test that was failing due to missalignment

ostannard reviewed (approved):
Thanks, LGTM

MaskRay commented on lld/test/ELF/aarch64-relocs.s:
```diff
@@ -30,7 +30,7 @@ mystr:
 # CHECK-EMPTY:
 # CHECK-NEXT: <.R_AARCH64_ADR_PREL_PG_HI21>:
-# CHECK-NEXT:   210132:       90000001        adrp    x1, 0x210000
+# CHECK-NEXT:   210134:       90000001        adrp    x1, 0x210000
```
the section addresses are insignificant. Just remove them.

popaflorin replied:
Thank you! I have merged now your changes

MaskRay commented on lld/test/ELF/aarch64-relocs.s:
```diff
@@ -231,24 +233,24 @@ movz1:
-# CHECK-NEXT: 2101c4: f2800021     movk	x1, #1
-# CHECK-NEXT: 2101c8: f29fffe1     movk	x1, #65535
-# CHECK-NEXT: 2101cc: d2a00042     mov	x2, #131072
+# CHECK-NEXT: 2101dc: d2800021     mov	x1, #1
```
These addresses are insignificant. I removed them in e1d1bb93d208c5772c385549c7efaa3d83459d2e

popaflorin replied:
Thank you! I have merged now your changes

MaskRay commented on llvm/test/MC/AArch64/directive-arch-section-alignment.s:
```diff
@@ -0,0 +1,32 @@
+// RUN: llvm-mc -triple=aarch64-none-linux-gnu -filetype=obj -o %t.obj %s
```
directive-arch isn't clear.

We don't need a new file. Reusing align.s should work

popaflorin replied:
I am unsure how to check for different alignment scenarios in the align.s. I thought the directive is a simple use case checking for various section alignment.

MaskRay replied:
Adding a new test is fine. Perhaps align-code.s

Please drop the non-meaningful triple aarch64-none-linux-gnu and switch to just `aarch64` (generic ELF), which is also short.

For ELF relocatable files, prefer `%t.o` to `%t.obj` (not COFF). Or just use `%t` in test/MC since almost all output files are ELF relocatable files.

popaflorin replied:
Thank you! I have done this changes in my last commit

popaflorin committed:
Merge branch 'main' into aarch64-section-alignment

flopop01 committed:
[AArch64][ELF] Renamed and updated the target for the test

Renamed the test to align-code.s and changed generic aarch64 ELF

Stylie777 commented:
I have merged this on behalf of @popaflorin, thanks for this!
